src/main/java/com/emilpiekos/usersmanager/controllers/RegisterController.java

```diff
@@ -0,0 +1,55 @@
+package com.emilpiekos.usersmanager.controllers;
+
+import com.emilpiekos.usersmanager.user.UserDto;
+import com.emilpiekos.usersmanager.user.UsersService;
+import jakarta.validation.ConstraintViolation;
+import jakarta.validation.ConstraintViolationException;
+import jakarta.validation.Valid;
+import org.springframework.stereotype.Controller;
+import org.springframework.ui.Model;
+import org.springframework.validation.BindingResult;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.ModelAttribute;
+import org.springframework.web.bind.annotation.PostMapping;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Controller
+public class RegisterController {
+
+    private final UsersService usersService;
+
+    public RegisterController(UsersService usersService) {
+        this.usersService = usersService;
+    }
+
+    @GetMapping("/register")
+    public String register(Model model) {
+        model.addAttribute("userDto", new UserDto());
+        return "register";
+    }
+
+    @PostMapping("/send-registration-form")
+    public String register(@Valid @ModelAttribute("userDto") UserDto userDto, BindingResult bindingResult, Model model) {
+        if (bindingResult.hasErrors()) {
+            return "register";
+        } else {
+            try {
+                usersService.save(userDto);
+                return "redirect:/register-success";
+            } catch (ConstraintViolationException e) {
+                Set<ConstraintViolation<?>> errors = e.getConstraintViolations();
+                Set<String> collect = errors.stream()
+                        .map(ConstraintViolation::getMessage).collect(Collectors.toSet());
+                collect.forEach(error -> model.addAttribute("error", error));
+                return "redirect:/error-page";
```

Comment:
Tutaj zamiast przekierowywać do strony błędu, lepiej byłoby wrócić użytkownika do strony z formularzem rejestracyjnym i wyświetlić mu błędy w formularzu, żeby mógł je od razu poprawić. Tak jak pisałem na slacku, aktualne rozwiązanie nie za bardzo zadziała, bo jak ustawisz atrybuty modelu i zrobisz redirect, to atrybuty te są tracone.
Należałoby użyć np. redirect attributes / flash attributes. No i tam ogólnie też mógłbyś zrobić tak, że zamiast przekierowywać do strony error-page, to mógłbyś spróbować w BindingResult ustawić złamane ograniczenia i też wyświetlić użytkownikowi formularz, czyli zamisat model.addAttribute to bindingResult.rejectValue()
https://stackoverflow.com/questions/12107503/adding-error-message-to-spring-3-databinder-for-custom-object-fields